### ISSUE TITLE
Future: Move from Rollup to Rolldown

### ISSUE THREAD

**Issue by birkskyum** [maplibre/maplibre-gl-js#4669]
We use rollup for bundling.

The vast majority of rollup use today happens as part of vite projects ([stats](https://npmtrends.com/rollup-vs-vite)).

The Vite team use rollup for prod builds, because of the [larger feature set](https://vitejs.dev/guide/why#why-not-bundle-with-esbuild) and flexible plugin system (supports bundle splitting and more), but rollup is really slow compared to a lot of alternatives, so for dev builds esbuild is used instead.

To simplify this setup, the [Rolldown](https://rolldown.rs/) projects is in development - it's a rust port of rollup, with compatible api/feature set, and performance superior to esbuild.

We can swap to Rollup with minimal changes when there's a stable release. It's [quite far](https://x.com/youyuxi/status/1824023061714399292) in development already - I've tested the latest release in this codebase, and it's getting there.

**Comment by HarelM** [comment]
I think we should evaluate this when it's a lot more stable and widely used. 
Last time we did a build change (I think it was swc) it ended up being rolled back due to issues we couldn't solve, and maybe only because we tried to use it prematurely.
We have a relatively complicated build (unfortunately) and we need this to work well - be battle tested. 
Just to clarify, I'm not against it, I just want to make sure we don't repeat past mistakes.

**Comment by birkskyum** [comment]
I fully agree, I've been thinking a lot about our options in this space and recently seen the tangible [performance gain in bundling](https://github.com/plotly/plotly.js/pull/6909#issuecomment-2304328740) (10-20x) and work required to move to esbuild, and I just wanted to offload my thoughts on the topic, because one option looked more feasible than others.

It would be neat to get a native bundler (like swc/esbuild), but only if it's low cost to migrate (which in our rullup-setup would be rolldown), and regarding stability I also think we should only seriously evaluate this after it becomes the Vite default, which is still some 1-1.5 years out, and maybe even after it overtakes rollup usage.

**TSC for type checking** isn't really possible to replace. I did the mistake in attempting this with SWC (didn't understand the project scope), which resulted in [the rollback](https://github.com/maplibre/maplibre-gl-js/pull/1399). There's one or two projects (not ready) still attempting to make a native type checker ([Ezno](https://github.com/kaleidawave/ezno)), but most have now admitted defeat ([SWC/STC](https://github.com/swc-project/swc/issues/571)), and there's now a lot of [resistance in attempting this feat](https://github.com/denoland/deno/issues/5432#issuecomment-2161434076).

**Comment by birkskyum** [comment]
We use AMD format currently in our Rollup config. Rolldown doesn't support that yet module format yet, so I've opened a tracking ticket:
- https://github.com/rolldown/rolldown/issues/2528#issuecomment-2450060073

Could be unblocked by
- #1595 
- #6254

**Comment by birkskyum** [comment]
Related to this topic
- https://github.com/vitejs/rolldown-vite/issues/585
